### ISSUE TITLE
Media Folder Management (File System Cleanup) taking more than 24 hours without response

### ISSUE THREAD

**Issue by TFAstudio** [FALM-Umbraco-Projects/FALM-Housekeeping-v7#41]
Hello

I have downloaded and installed your FALM Housekeeping v7.6.0.7 that is needed in previous versions of Umbraco 7.x.  It loads.... so thank you for testing that out and getting it to work.

We have a media folder that is loooooong overdue for something like this, with several GB of images.  I installed the FALM and activated it for an administrative user, then clicked on the tab, opened the Media section and clicked on File System Cleanup.

It gave me the following screen:
![image](https://user-images.githubusercontent.com/19319418/33865289-2ed88808-df45-11e7-8a0b-5fcb70ef52d9.png)

The triple dots just keep moving and nothing happens and it has been quite some time - what is the time I should expect and is there any way to find out the status of how far things have gone?

Thanks

**Comment by TFAstudio** [comment]
Still going - nearly 36 hours later...  Is this expected?

**Comment by afabri73** [comment]
Hello, no...this is too much.
I tried in a website with a lot of media but I haven't wait more than 2 hours.
Surely there has been something that gone wrong. (sorry my not perfect english)

I don't know how many media nodes, folders and files you have in your website...but I think 36 hours is too much.
However...you must think that the media manager do some things before return the search result:
1 - Check all folders and files in filesystem inside media folder and compare each one with all media entries in the Umbraco DB (this is a long procedure)
2 - Check all media entries in the Umbraco DB and compare each one with all folders and files in filesystem inside media folder  (also this procedure is long).

It may seem that this is a useless double check, but I assure that it is not.
This gives the absolute certainty that the files and folders to be removed are actually orphans thus avoiding errors
The bad thing is that unfortunately the execution times are double.

I tried to remove a passage, in the past, but the cleaning was not complete

**Comment by TFAstudio** [comment]
Is there anything I can check to see what may be going wrong?  Any logs or files that are written out to help with debugging?

Thanks for the response!

**Comment by TFAstudio** [comment]
Just looking through a sql profiler on my local version - I notice you are using a like statement with a wildcard at the beginning and the end - and (I am sure you do not need me to tell you that) this kind of statement has notoriously bad performance.  I notice that you are repeatedly hitting the `cmsContentXml` and `cmsPropertyData` tables checking to see if there are IDs.  Often I notice that the `%/media/<NUMBER>/%` folder you are testing, the `<NUMBER>` does not change through the subsequent queries....  as if you have not cached the results of your query?  This seems to be your `GetCountFromCmsContentXml()` and `GetCountFromCmsPropertyData()` methods

I am curious why you do not cache the results in your `foreach` loop(s) so you do not have to hit the database again - this query will return the same result time and again (well, if you discount new media entries that may be happening behind the scenes which will probably not need to be deleted)

I have attempted a small change to the `HKMediaApiController.cs` and include it here as a patch that tries to cache the results.  I have not tested it, it is just an idea/suggestion that perhaps you can try?

```
--- D:\Test\HKMediaApiController.cs.orig	2017-12-15 09:29:59.000000000 +1100
+++ D:\Test\HKMediaApiController.cs	2017-12-15 09:44:43.000000000 +1100
@@ -39,12 +39,14 @@
         /// <returns>MediaModel</returns>
         [HttpGet]
         public HKMediaModel GetMediaToDelete(string userLocale)
         {
             ListMediaWarnings = new List<MediaWarningModel>();
             ListMediaToDelete = new List<MediaToDeleteModel>();
+            // Map of media folders to the count of information within them to minimize database calls
+            Dictionary<string, int> mediaCountMap = new Dictionary<string, int>();
 
             var currentUserCultureInfo = CultureInfo.GetCultureInfo(userLocale);
 
             try
             {
                 // Find all media to be deleted and relative warning messages
@@ -95,12 +97,14 @@
                             if (int.TryParse(d1.Name, out n1) && int.TryParse(d2.Name, out n2))
                             {
                                 return n1 - n2;
                             }
                             return string.CompareOrdinal(d1.Name, d2.Name);
                         });
+                        
+                        int mediaCount = -1;
 
                         foreach (var subDir in subDirs)
                         {
                             // Do check only if the folder have a number as a name (STANDARD FOLDER)
                             MediaWarningModel mediaWarningModel;
 
@@ -125,18 +129,18 @@
                                                     mediaToSkip.Add(subDir.Name);
                                                     break;
                                                 }
                                                 if (iDirectoryName < pId)
                                                 {
                                                     // Check if the folder is used by data type that not store image informations (like Image Cropper)
-                                                    int mediaCount = 0;
-
-                                                    mediaCount += GetCountFromCmsContentXml(db, subDir.Name);
-
-                                                    mediaCount += GetCountFromCmsPropertyData(db, subDir.Name);
-
+                                                    if (!mediaCountMap.ContainsKey(subDir.Name))
+                                                    {
+                                                        mediaCountMap[subDir.Name] = GetCountFromCmsContentXml(db, subDir.Name) + GetCountFromCmsPropertyData(db, subDir.Name);
+                                                    }
+                                                    mediaCount = mediaCountMap[subDir.Name];
+=
                                                     // The Media is deletable if it isn't used
                                                     if ((mediaCount == 0) && (!mediaAlreadyAddedToDeleteList))
                                                     {
                                                         // MEDIA FOUND IN FILE SYSTEM BUT NO CORRISPONDATION INTO DB OR BY DATATYPE UPLOAD
                                                         // ### DELETEBLE ###
                                                         mediaToDeleteModel = new MediaToDeleteModel
@@ -179,17 +183,17 @@
 
                                 }
 
                                 if (!mediaToSkip.Contains(subDir.Name))
                                 {
                                     // Check if the folder is used by data type that not store image informations (like Image Cropper)
-                                    var mediaCount = 0;
-
-                                    mediaCount += GetCountFromCmsContentXml(db, subDir.Name);
-
-                                    mediaCount += GetCountFromCmsPropertyData(db, subDir.Name);
+                                    if (!mediaCountMap.ContainsKey(subDir.Name))
+                                    {
+                                        mediaCountMap[subDir.Name] = GetCountFromCmsContentXml(db, subDir.Name) + GetCountFromCmsPropertyData(db, subDir.Name);
+                                    }
+                                    mediaCount = mediaCountMap[subDir.Name];
 
                                     // If the media is not used...it is deletable
                                     if (mediaCount == 0)
                                     {
                                         // MEDIA FOUND IN FILE SYSTEM BUT NO CORRISPONDATION INTO DB OR BY DATATYPE UPLOAD
                                         // ### DELETEBLE ###

```

**Comment by TFAstudio** [comment]
I have tested this and what was not completing for _several days_ was completed within 20-30 minutes with the changes above.  I will clean up the code to ensure there is proper garbage collection.  There is a slight trade-off between memory and speed/database calls, but with the over 5GB of media, the memory was fairly negligible.

I have tested this on the earlier release of FALM, so I cannot verify that this makes an appropriate difference on later versions, but I cannot see why it would not.
